File: packages/health_connector_hc_android/lib/src/mappers/health_record_mappers/respiratory_rate_record_mapper.dart
import 'package:health_connector_core/health_connector_core_internal.dart'
    show RespiratoryRateRecord, HealthRecordId, sinceV1_3_0, Frequency;
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/health_record_id_mapper.dart';
import 'package:health_connector_hc_android/src/mappers/metadata_mappers/metadata_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart';
import 'package:meta/meta.dart' show internal;

/// Mapper extensions for [RespiratoryRateRecord].
@sinceV1_3_0
@internal
extension RespiratoryRateRecordToDto on RespiratoryRateRecord {
  /// Converts [RespiratoryRateRecord] to its [RespiratoryRateRecordDto].
  RespiratoryRateRecordDto toDto() {
    return RespiratoryRateRecordDto(
      id: id.toDto(),
      time: time.toUtc().millisecondsSinceEpoch,
      zoneOffsetSeconds: zoneOffsetSeconds,
      metadata: metadata.toDto(),
      breathsPerMinute: rate.inPerMinute,
    );
  }
}

/// Mapper extensions for [RespiratoryRateRecordDto].
@sinceV1_3_0
@internal
extension RespiratoryRateRecordDtoToDomain on RespiratoryRateRecordDto {
  /// Converts [RespiratoryRateRecordDto] to its domain representation.
  RespiratoryRateRecord toDomain() {
    return RespiratoryRateRecord.internal(
      id: id != null ? HealthRecordId(id!) : HealthRecordId.none,
      time: DateTime.fromMillisecondsSinceEpoch(time, isUtc: true),
      zoneOffsetSeconds: zoneOffsetSeconds,
      metadata: metadata.toDomain(),
      rate: Frequency.perMinute(breathsPerMinute),
    );
  }
}

File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/RespiratoryRateRecordMapper.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.RespiratoryRateRecord
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.RespiratoryRateRecordDto
import java.time.Instant
import java.time.ZoneOffset

/**
 * Converts a Health Connect [RespiratoryRateRecord] to a [RespiratoryRateRecordDto].
 */
internal fun RespiratoryRateRecord.toDto(): RespiratoryRateRecordDto = RespiratoryRateRecordDto(
    id = metadata.id,
    time = time.toEpochMilli(),
    zoneOffsetSeconds = zoneOffset?.totalSeconds?.toLong(),
    metadata = metadata.toDto(),
    breathsPerMinute = rate,
)

/**
 * Converts a [RespiratoryRateRecordDto] to a Health Connect [RespiratoryRateRecord].
 */
internal fun RespiratoryRateRecordDto.toHealthConnect(): RespiratoryRateRecord =
    RespiratoryRateRecord(
        rate = breathsPerMinute,
        time = Instant.ofEpochMilli(time),
        zoneOffset = zoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) },
        metadata = metadata.toHealthConnect(id),
    )
